### PR TITLE
GEOMESA-1404 Classpath Additions to MapReduce Ingest

### PR DESCRIPTION
**Pull request by aheyne**
Add the ability to add items to class path by using the system variable geomesa.convert.scripts.path

Signed-off-by: Austin Heyne <aheyne@ccri.com>

### CODE REVIEW

**Comment by aheyne** [comment]
This code is a little out of my current knowledge, critique welcome.

File: geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngestJob.scala [hunk 1]
```diff
@@ -96,6 +96,7 @@ abstract class AbstractIngestJob extends LazyLogging {
   def ingestJarSearchPath: Iterator[() => Seq[File]] =
     Iterator(() => ClassPathUtils.getJarsFromEnvironment("GEOMESA_HOME"),
       () => ClassPathUtils.getJarsFromEnvironment("ACCUMULO_HOME"),
+      () => ClassPathUtils.getJarsFromSystemProperty("geomesa.convert.scripts.path"),
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngestJob.scala]: Does this need to be documented somehow, somewhere?

Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngestJob.scala]: +1

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala [hunk 2]
```diff
@@ -80,6 +80,21 @@ object ClassPathUtils extends LazyLogging {
...
+    * Finds URLs of jar files based on a system property
+    *
+    * @param prop
+    * @return
+    */
+  def getJarsFromSystemProperty(prop: String): Seq[File] = {
+    Option(System.getProperty(prop)) match {
+      case Some(path) => path.toString().split(":").map(new File(_)).filter(_.isDirectory).toSeq.flatMap(loadJarsFromFolder)
+      case None =>
+        logger.warn(s"Can't load jars from System Property: ${prop}")
```
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: i wouldn't use a logger.warn here - this is probably the normal, default case. might want to add some warnings in the `filter(_.isDirectory)` check instead.

**Comment by jnh5y** [comment]
@elahrvivaz can you re-review and see if your issues have been addressed?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala [hunk 3]
```diff
@@ -98,6 +98,37 @@ object ClassPathUtils extends LazyLogging {
...
+  /**
+    * Recursively searches folders for all files
+    *
+    * @param dir
+    * @return
+    */
+  def loadFilesFromFolder(dir: File): Seq[File] = {
+    val files = Option(dir.listFiles()).toSeq.flatten
+    val children = Option(dir.listFiles(folderFileFilter)).toSeq.flatten.flatMap(loadFilesFromFolder)
+    files ++ children
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: Does this lead to any duplication?  Does a sub-directory get listed twice ever?  Does that even matter;)?

**Comment by jnh5y** [comment]
@elahrvivaz I'm assigning this one to you to push-da-button.

**Comment by elahrvivaz** [comment]
@aheyne have you tested this out with an ingestion?

**Comment by aheyne** [comment]
I've tested the code works in repl but I haven't through an ingest, I can try to set that up later.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala (hunk 3 above)
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: you can add a filter to select files vs folders in the list() method on a file or filter with .filter(_.isFile) or something like that

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala [hunk 4]
```diff
@@ -98,6 +98,37 @@ object ClassPathUtils extends LazyLogging {
...
+    *
+    * @param prop
+    * @return
+    */
+  def getFilesFromSystemProperty(prop: String): Seq[File] = {
+    Option(System.getProperty(prop)) match {
+      case Some(path) => path.toString().split(":").map(new File(_)).filter{ file =>
+        val isDir = file.isDirectory
+        if (!isDir) logger.warn(s"${file} : Is not a directory")
+        isDir
```
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: what if i want to add a single file? can we support that?

Comment (aheyne) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: Added support for single files

**Comment by jahhulbert-ccri** [comment]
has this been tried with an example convert script? e.g. a bit of javascript that needs to be added to a mapreduce job for ingest?

**Comment by jnh5y** [comment]
@jahhulbert-ccri that's a good question.  Do we have an easy example of such a job?

**Comment by aheyne** [comment]
Easy example would be great because I'm having trouble figuring out how to test this on my own.

**Comment by jahhulbert-ccri** [comment]
@aheyne I think the idea would be to just have a cloud local and use a java script convert function to modify an attribute i'll try to wip up an example

**Pull request merged**